share/spack/gitlab/cloud_pipelines/stacks/ml-linux-power64le-cpu/spack.yaml

```diff
@@ -0,0 +1,142 @@
+      variants: ~cuda~rocm
+#    pkgconfig:
+#      require: pkgconf
+#    pil:
+#      require: py-pillow
+#    mpi:
+#      require: openmpi
+#    python:
+#      require: "+libxml2"
+#    py-pybind11:
+#      require: "build_system=cmake"
+#    py-keras-applications:
+#      require: "build_system=python_pip"
+#    curl:
+#      require: "tls=gnutls"
+#    py-pytorch-gradual-warmup-lr:
+#      require: "build_system=python_pip"
+#    py-pytorch-lightning:
+#      require: "build_system=python_pip"
+#    jpeg:
+#      require: "libjpeg-turbo"
+#    py-torch-cluster:
+#      require: "~cuda"
+#    py-torch-geometric:
+#      require: "~cuda"
+#    py-torch-scatter:
+#      require: "~cuda"
+#    py-torch-spline-conv:
+#      require: "~cuda"
+#    py-torchaudio:
+#      require: "build_system=python_pip"
+#    py-torchfile:
+#      require: "build_system=python_pip"
+#    py-torchvision:
+#      require: "build_system=python_pip"
+#    py-numpy:
+#      require: "build_system=python_pip"
+#    py-tensorboard-data-server:
+#      require: "build_system=python_pip"
+#    py-tensorboard-plugin-wit:
+#      require: "build_system=python_pip"
+#    swig:
+#      require: "@4.1.1"
+#    xgboost:
+#      require: "build_system=cmake generator=ninja"
+
+  specs:
+    # Horovod
+    - py-horovod tensor_ops=mpi controllers=mpi frameworks=pytorch
+
+    # Hugging Face
+    - py-transformers
+
+    # JAX
+    - py-jax
+    - py-jaxlib
+
+    # Keras
+    - py-keras
+    - py-keras-applications
+    - py-keras-preprocessing
+    - py-keras2onnx
+
+    # MXNet
+    - mxnet
+
+    # PyTorch
+    - py-botorch
+    - py-efficientnet-pytorch
+    - py-gpytorch
+    - py-kornia
+    - py-lightning
+    - py-pytorch-gradual-warmup-lr
+    - py-pytorch-lightning
+    - py-segmentation-models-pytorch
+    - py-timm
+    - py-torch
+    - py-torch-cluster
+    - py-torch-geometric
+    - py-torch-nvidia-apex
+    - py-torch-scatter
+    - py-torch-sparse
+    - py-torch-spline-conv
+    - py-torchaudio
+    - py-torchdata
+    - py-torchfile
+    - py-torchgeo
+    - py-torchmetrics
+    - py-torchtext
+    - py-torchvision
+    - py-vector-quantize-pytorch
+
+    # scikit-learn
+    - py-scikit-learn
+    - py-scikit-learn-extra
+
+    # TensorBoard
+    - py-tensorboard
+    - py-tensorboard-data-server
+    - py-tensorboard-plugin-wit
+    - py-tensorboardx
+
+    # TensorFlow
+    - py-tensorflow
+    - py-tensorflow-datasets
+    - py-tensorflow-estimator
+    - py-tensorflow-hub
+    - py-tensorflow-metadata
+    - py-tensorflow-probability
+
+    # XGBoost
+    - py-xgboost
+    # - r-xgboost
+    - xgboost
+
+  mirrors:
+    mirror: s3://spack-binaries/develop/ml-linux-power64le-cpu
+
+  ci:
+    pipeline-gen:
+    - build-job:
+        image:
+          name: ecpe4s/ubuntu20.04-runner-ppc64le:2023-01-01
+          entrypoint: ['']
+    - build-job:
+        image:
+          name: ecpe4s/fedora36-runner-ppc64le:2023-01-01
+          entrypoint: ['']
```

Comment:
I think one of these `build-job` sections should be removed.  @kwryankrattiger can correct me when I get this backwards, but I think these are applied bottom up, so your final job configuration will just have the first one.  In which case, you can remove the second.

Reply:
So, I assume you mean only have one of the ubuntu or fedora jobs.  I'm actually intentionally targeting both as many of these packages, if they even support linux, are built for Ubuntu.  If users are installing these packages locally on a linux system, it is very likely Ubuntu.  But most of the HPC systems that I've interacted with are either SLES or RHEL-based.  As such, I think it is useful to ensure that these packages are tested against both OSes as best we can.

If I'm just specifying it incorrectly to achieve that end, then I'm happy to change it.

Reply:
I think testing both ubuntu and rhel is overkill right now. We don't even do that for x86_64. This is something worth adding someday but idk if we have the CI bandwidth to do both yet.

Reply:
I think testing both ubuntu and rhel is overkill right now. We don't even do that for x86_64. This is something worth adding someday but idk if we have the CI bandwidth to do both yet.

Reply:
I've already experienced builds that succeeded with ubuntu but not with rhel8.  I'm leaving it in because tests that pass on ubuntu do us no good when the majority of HPC systems are on a rhel-based OS.  It really comes down to what our goal is with the CI system.  If we want to ensure that changes to these packages won't have a significant impact to most of our users that try and build these recipes on most of the systems they might use, then we need to ensure that our CI is representative of those systems and users.  If we just want a build that works even if it doesn't represent anything an end user might use, we can probably just ignore ppc64le altogether.  As for CI bandwidth, most changes in Spack shouldn't even trigger this.  The point (as I see it) is to test for any changes that will impact these builds.  Once these pipelines are integrated, they shouldn't be triggered often, unless they need to be triggered.  Not sure how we judge this bandwidth or if such limiting factors are publicly available, but I think this is necessary for now.  Obviously, once all of these systems have been decommissioned, we can remove these pipelines.  Until then, these pipelines are a useful tool for trying to get these ML packages working on the machines we want them to build on.